src/scheduler/pbs_sched.c

```diff
@@ -1279,84 +1185,49 @@ main(int argc, char *argv[])
 
 	pthread_mutex_init(&cleanup_lock, &attr);
 
-	FD_ZERO(&fdset);
+	connect_servers();
 
 	for (go=1; go;) {
-		int	cmd;
+		svr_t *sconn = NULL;
 
-		FD_SET(server_sock, &fdset);
-		if (select(FD_SETSIZE, &fdset, NULL, NULL, NULL) == -1) {
-			if (errno != EINTR) {
-				log_err(errno, __func__, "select");
-				die(0);
-			}
-			continue;
-		}
+		wait_for_cmd(&sconn);
 
 #ifdef PBS_UNDOLR_ENABLED
 		if (sigusr1_flag)
 			undolr();
 #endif
-		if (!FD_ISSET(server_sock, &fdset))
-			continue;
-
-		/* connector is set in server_connect() */
-		cmd = server_command(&runjobid);
 
-		if (connector >= 0) {
-			if (update_svr) {
-				/* update sched object attributes on server */
-				update_svr_schedobj(connector, cmd, alarm_time);
-				update_svr = 0;
-			}
-
-			if (sigprocmask(SIG_BLOCK, &allsigs, &oldsigs) == -1)
-				log_err(errno, __func__, "sigprocmask(SIG_BLOCK)");
-
-			/* Keep track of time to use in SIGSEGV handler */
-#ifdef NAS /* localmod 031 */
-			now = time(NULL);
-			if (!opt_no_restart)
-				segv_last_time = now;
-			{
-				strftime(log_buffer, sizeof(log_buffer),
-					"%Y-%m-%d %H:%M:%S", localtime(&now));
-				printf("%s Scheduler received command %d\n", log_buffer, cmd);
-			}
-#else
-			if (!opt_no_restart)
-				segv_last_time = time(NULL);
-
-			DBPRT(("Scheduler received command %d\n", cmd));
-#endif /* localmod 031 */
-
-			if (schedule(cmd, connector, runjobid)) /* magic happens here */ {
-				go = 0;
-			}
-			if (second_connection != -1) {
-				close(second_connection);
-				second_connection = -1;
-			}
-
-			if (server_disconnect(connector))
-				connector = -1;
-
-			if (runjobid != NULL) {
-				free(runjobid);
-				runjobid = NULL;
-			}
+		if (sigprocmask(SIG_BLOCK, &allsigs, &oldsigs) == -1)
+			log_err(errno, __func__, "sigprocmask(SIG_BLOCK)");
+
+		/* Keep track of time to use in SIGSEGV handler */
+		now = time(NULL);
+		if (!opt_no_restart)
+			segv_last_time = now;
+#ifdef DEBUG
+		{
+			strftime(log_buffer, sizeof(log_buffer), "%Y-%m-%d %H:%M:%S", localtime(&now));
+			DBPRT(("%s Scheduler received command %d\n", log_buffer, sconn->cmd));
+		}
+#endif
 
-			if (sigprocmask(SIG_SETMASK, &oldsigs, NULL) == -1)
-				log_err(errno, __func__, "sigprocmask(SIG_SETMASK)");
+		if (schedule(sconn)) /* magic happens here */ {
+			go = 0;
+		} else {
```

Comment:
Why an else?  Don't you want to do a schedule() and then a send_end_cycle() right after it?

Reply:
if schedule() returns !0 then will be go out of this loop, and very next thing we do is close server connections, which can lead server to read on closed fd, aka server will get EOF not what sched sent, so no in cause if we are going out of loop we don't have to send end cycle notification as server will get notified by closed connection.